### PR TITLE
Docker compose setup

### PR DESCRIPTION
**Pull request by benryder1988**
**Story card:** [ch1907](https://app.clubhouse.io/simpledotorg/story/1907/docker-setup)

## Because

Local setup of simple-server could be simplified and capable of running in a Docker Compose environment to enable developers to rapidly spin up the server to test android app changes etc. 

## This addresses

Adding docker-compose support to quickly deploy simple-server locally

### CODE REVIEW

**Comment by vinaysshenoy** [comment]
Sweet! This was a major blocker for me never running the server locally to test changes on the mobile app.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
+FROM ruby:2.5.1
+
+RUN curl https://deb.nodesource.com/setup_12.x | bash
+RUN curl https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add -
+RUN echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list
+
+RUN apt-get update && apt-get install -y yarn redis-server postgresql-client nodejs jq
+
+# throw errors if Gemfile has been modified since Gemfile.lock
+RUN bundle config --global frozen 1
```
Comment (kitallis) [inline Dockerfile:10]: This is neat.

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      - "6379"
+    volumes:
+      - redisdata:/data
+
+  server:
+    build: .
+    volumes:
+      - .:/myapp
+    command: >
+      sh -c "sleep 10 &&
```
Comment (kitallis) [inline docker-compose.yml:41]: The `sleep` is for?

**Review by kitallis — COMMENTED** [review]
Thanks so much @benryder1988! 

This will definitely come in very handy. I'll come back for another pass once I have chance to test it fully.

File: docker-compose.yml (hunk 2 above)
Comment (benryder1988) [reply docker-compose.yml:41]: I realised that the at least one of the services that the server depends on (I think postgres) takes a few seconds to start fully once the containers spin up, but the 'depends on' docker compose feature is satisfied as the service is technically running. The sleep basically pauses the server trying to connect until all the other service should be ready - but I admit it is a bit of a hack :)

**Comment by benryder1988** [comment]
Hi all, you're welcome and happy to help - glad to hear the feature should be useful! 

Just FYI:  When I making these changes a few weeks ago, I noticed that the 'pg_running' method in bin/setup would fail as postgres wasn't running locally in the container. But when I tested again yesterday the issue seemed to be gone. Let me know if you run into the same problem - I have an idea of how to possibly fix it.

**Comment by harimohanraj89** [comment]
@benryder1988 we took this for a spin and it looks great! We were able to run the server smoothly with and without docker. Didn't run into the `bin/setup` problem you were describing, so no worries.

Our CI isn't configured to run against forks yet. I'll set it up, ensure the build's green, and merge this in later today. Thanks for setting this up!

**Comment by benryder1988** [comment]
Hi @harimohanraj89, great news and thanks for the update! Happy to help and looking forward to contributing more in the future :)

File: docker-compose.yml (hunk 2 above)
Comment (kitallis) [reply docker-compose.yml:41]: No problem, these kind of things are acceptable in setup scripts I feel :)

**Review by harimohanraj89 — APPROVED** [review]
Getting Semaphore to re-run specs on this PR was clunky. I verified the build is green in a cloned branch: https://resolvetosavelives.semaphoreci.com/workflows/637f68bf-8d0e-4d05-8b94-5e950b17f997?pipeline_id=3d92c1fb-f257-4a6f-8246-410e8ef9e9a2

:shipit:

**Pull request merged**